File: Common/src/message/Messageable.java
package message;

import java.io.InputStream;
import java.io.OutputStream;

/**
 *
 * @author Alberto Pareja-Lecaros
 *
 * Defines methods that clients and servers must implement in order to 
 * interpret messages received through a MessageHandler.
 *  
 */
public interface Messageable {

	/**
	 * 
	 * Determines a course of action based on the code passed and the state of
	 * the Messageable.
	 * 
	 * @param code - the integer code received from a network source
	 * @return None
	 */
	public void interpretCode(Message message);
	
	public int getPort();
	
	public String getHost();
	
	public InputStream getInputStream();
	
	public OutputStream getOutputStream();
	
	public InputStream getFileInputStream();
	
	public OutputStream getFileOutputStream();
	
	public boolean isConnected();
	
	public boolean isLoggedIn();
	
	public void receiverDisconnected();
	
	public String getDeviceName();
	
	public String getOsName();
	
	public String getRootFolder();
	
	public void fileReceivedCallback(String file, Message m);
}

File: Common/src/main/java/message/Messageable.scala
package message;

import java.io._
import java.net._

/**
 * Defines methods that clients and servers must implement in order to 
 * interpret messages received through a MessageHandler.
 */
trait Messageable {
  
  var host: String
  var port: Int
  var connected: Boolean
  var loggedIn: Boolean
  var osName: String
  var rootFolder: String
  var socket: Socket
  var fileSocket: Socket
  var messageSender: MessageSender

  /**
   * 
   * Determines a course of action based on the code passed and the state of
   * the Messageable.
   * 
   * @param code - the integer code received from a network source
   * @return None
   */
  def interpretCode(message: Message): Unit
  def receiverDisconnected: Unit
  def fileReceivedCallback(filename: String, message: Message): Unit
  
  def inputStream: InputStream = {
    try {
      return socket.getInputStream
    } catch {
      case e: IOException => e.printStackTrace
    }
	null
  }
  def outputStream: OutputStream = {
    try {
      return socket.getOutputStream
    } catch {
      case e: IOException => e.printStackTrace
    }
	null
  }
  
  def fileInputStream = {
    try {
      fileSocket.getInputStream
    } catch {
      case e: IOException => e.printStackTrace
    }
    null
  }
  
  def fileOutputStream = {
    try {
      fileSocket.getOutputStream
    } catch {
      case e: IOException => e.printStackTrace
    }
    null
  }
  
  def deviceName: String
}
